tests/unit/s2n_server_hello_test.c

```diff
@@ -468,6 +468,52 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_connection_free(server_conn));
     }
 
+    /* Verify a TLS1.1 client cannot negotiate with a default server but then
+     * updated to TLS1.2 after the s2n_config_set_min_protocol_version() call */
+    {
+        struct s2n_config *client_config;
+        struct s2n_connection *client_conn;
+        struct s2n_config *server_config;
+        struct s2n_connection *server_conn;
+
+        EXPECT_NOT_NULL(server_config = s2n_config_new());
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_NOT_NULL(client_config = s2n_config_new());
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, client_config));
+
+        EXPECT_SUCCESS(s2n_config_set_min_protocol_version(server_config, S2N_TLS12));
+
+        /* The client will request TLS1.1 */
+        client_conn->client_protocol_version = S2N_TLS11;
+
+        struct s2n_stuffer *server_stuffer = &server_conn->handshake.io;
+
+        const uint32_t total = S2N_TLS_PROTOCOL_VERSION_LEN
+            + S2N_TLS_RANDOM_DATA_LEN
+            + SESSION_ID_SIZE
+            + server_conn->session_id_len
+            + S2N_TLS_CIPHER_SUITE_LEN
+            + COMPRESSION_METHOD_SIZE;
+
+        server_conn->secure.cipher_suite = &s2n_ecdhe_ecdsa_with_aes_128_gcm_sha256;
+        EXPECT_SUCCESS(s2n_server_hello_send(server_conn));
+        EXPECT_EQUAL(s2n_stuffer_data_available(server_stuffer), total);
+
+        /* Copy server stuffer to client stuffer */
+        EXPECT_SUCCESS(s2n_stuffer_copy(&server_conn->handshake.io, &client_conn->handshake.io, total));
+
+        /* Verify that client errors due to lower version */
+        EXPECT_FAILURE(s2n_server_hello_recv(client_conn));
```

Comment:
Can you check for specific error codes using `EXPECT_FAILURE_WITH_ERRNO()`? I'd like to see that these calls are failing for the reasons we expect.

Reply:
Not just "allow TLS1.3", but also allow "TLS 1.1 and above", or "TLS 1.2 and above", basically setting the base version bar for connections. We already have that mechanism "minimum_protocol_version" but it is now coupled with cipher preferences.

Currently open source Redis, depending on Openssl, has this functionality for customers and we are just looking for parity.

I talked with Alex and he proposed me that I could either add ElastiCache specific cipher preferences with "minimum_protocol_version" as 1.0, 1.1, 1.2 and 1.3 each, or add this new API to generically solve this problem. He recommended this new API solution and I agreed with him.

Reply:
Am I missing the `s2n_connection_set_minimum_supported_version()` function? Or do you only need this done on a configuration?

Not sure if I'm missing any updates, can you verify the failure code here? We know s2n_server_hello_recv fails, but I don't know why. We will want to make sure the error stays consistent.